# WEEK 2

## Lenear regression in gapminder.csv

### Importing libraries

In [1]:
import pandas as pd
import numpy as np
import statsmodels.formula.api as sm

### Importing dataset

In [2]:
data = pd.read_csv("gapminder.csv")

In [4]:
data1 = data.iloc[:,[0,1,2,9,14]].copy()

In [5]:
data1.head()

country   incomeperperson alcconsumption lifeexpectancy  \
0  Afghanistan                              .03         48.673   
1      Albania  1914.99655094922           7.29         76.918   
2      Algeria  2231.99333515006            .69         73.131   
3      Andorra  21943.3398976022          10.17                  
4       Angola  1381.00426770244           5.57         51.093   

         employrate  
0  55.7000007629394  
1  51.4000015258789  
2              50.5  
3                    
4  75.6999969482422

In [6]:
data1.columns = ["country","income","alcohol","life","eployment"]

### converting to float

In [8]:
def fun(a):
    
    for i in range(len(a)):
        if a[i] == " ":
            a[i]  = np.nan
        else :
            a[i] = float(a[i])
            

In [10]:
for i in data1.columns[1:]:
    fun(data1[i])

In [11]:
data1.head()

country   income alcohol    life eployment
0  Afghanistan      NaN    0.03  48.673      55.7
1      Albania     1915    7.29  76.918      51.4
2      Algeria  2231.99    0.69  73.131      50.5
3      Andorra  21943.3   10.17     NaN       NaN
4       Angola     1381    5.57  51.093      75.7

### missing values filled with mean

In [12]:
def fun1(a):
    
    x = np.mean(a.dropna())
    for i in range(len(a)):
        
        if np.isnan(a[i]):
            a[i] = x

In [13]:
for i in data1.columns[1:]:
    fun1(data1[i])

In [14]:
data1.head()

country   income alcohol     life eployment
0  Afghanistan  8740.97    0.03   48.673      55.7
1      Albania     1915    7.29   76.918      51.4
2      Algeria  2231.99    0.69   73.131      50.5
3      Andorra  21943.3   10.17  69.7535    58.636
4       Angola     1381    5.57   51.093      75.7

## Lenear regression income vs alcohol

### taking care of outliers (removing any data more than 4 standard deviation from the mean)

In [55]:
data2 = data1.iloc[:,[1,2]].copy()
data2.head()

income alcohol
0  8740.97    0.03
1     1915    7.29
2  2231.99    0.69
3  21943.3   10.17
4     1381    5.57

In [56]:
def fun2(a):
    
    mean = np.mean(a)
    std = np.std(a)
    mx = mean + 4*std
    mn = mean - 4*std
    
    for i in range(len(a)):
        if (a[i]>mx)|(a[i]<mn):
            a[i] = np.nan
        

In [57]:
for i in data2.columns:
    fun2(data2[i])

In [58]:
len(data2)

213

In [59]:
data2 = data2.dropna()

In [60]:
len(data2)

210

In [61]:
data2["alcohol"] = pd.to_numeric(data2["alcohol"])

In [62]:
data2["income"] = pd.to_numeric(data2["income"])

### 3 rows removed as they had outliers as values

# Final model

In [63]:
x = np.mean(data2["alcohol"])

In [64]:
y = np.mean(data2["income"])


In [65]:
y

7677.852807190078

## Centering done

In [66]:
for i in range(len(data2)):
    
    data2.iloc[i,0] =  data2.iloc[i,0]-y
    

data2.head()
    

income  alcohol
0   1063.113269     0.03
1  -5762.856256     7.29
2  -5445.859472     0.69
3  14265.487090    10.17
4  -6296.848539     5.57

In [67]:
model =  sm.ols(formula = "alcohol~income" , data = data2).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                alcohol   R-squared:                       0.071
Model:                            OLS   Adj. R-squared:                  0.067
Method:                 Least Squares   F-statistic:                     15.92
Date:                Tue, 28 Jul 2020   Prob (F-statistic):           9.18e-05
Time:                        15:01:13   Log-Likelihood:                -611.22
No. Observations:                 210   AIC:                             1226.
Df Residuals:                     208   BIC:                             1233.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.6894      0.308     21.708      0.0

# SUMMARY

As you can see in the summary F static is quite high and P value is very small so that means alchohol is related to income . Although since r^2 value is .071 shows that it income only explains 7.1% of the variablity and the lenear line we get is -

                alchohol = (0.0001)*income + (6.69)
here income is centered around mean 7677